### PR TITLE
Adding extra information about invariant messages

### PR DESCRIPTION
**Pull request by thk123**
Saw a couple of PRs getting comments relating to the grammatical style of invariant messages. I've added a clarification to the coding standard which reflects what I do and have seen others do so hopefully this isn't contentious!

### CODE REVIEW

File: CODING_STANDARD.md [hunk 1]
```diff
@@ -188,6 +188,10 @@ Formatting is enforced using clang-format. For more information about this, see
 - Avoid `assert`. If the condition is an actual invariant, use INVARIANT,
   PRECONDITION, POSTCONDITION, CHECK_RETURN, UNREACHABLE or DATA_INVARIANT. If
   there are possible reasons why it might fail, throw an exception.
+    - Use "should" style statements for messages in invariants (e.g. "Array
```
Comment (tautschnig) [inline CODING_STANDARD.md]: While at it: should those messages start with a lowercase or an uppercase character?

File: CODING_STANDARD.md [hunk 2]
```diff
@@ -188,6 +188,10 @@ Formatting is enforced using clang-format. For more information about this, see
...
+      should have a non-zero size") to make it clear both the violation and the
```
Comment (tautschnig) [inline CODING_STANDARD.md]: I think it should be "to make clear both the violation ..." (drop "it")

**Review by tautschnig — APPROVED** [review]
Comments on some further refinement below, but guidance of this sorts is very useful!

**Comment by danpoe** [comment]
This seems to slightly clash with the documentation in `invariant.cpp` which says that the message should state the reason for why it holds. For example something like `"array has non-zero size as it was checked by the caller"`. Whatever style we choose, we should probably update the documentation in `invariant.cpp` as well.

**Comment by thk123** [comment]
Sorry lost the will on this one! Is the verdict  `invariant.cpp` should be updated to match the wording proposed in this PR? I believe the wording I proposed more closely matches how we use them: to document coders assumptions rather than the `invariant.cpp` wording which hopes for precisely known invariants.

**Comment by tautschnig** [comment]
I think the wording in `invariant.cpp` should be updated, because aspects like "was checked by the caller" are expressed by using a `PRECONDITION` (while a `DATA_INVARIANT` should hold by construction).

File: CODING_STANDARD.md (hunk 1 above)
Comment (thk123) [reply CODING_STANDARD.md]: 72 upper case vs. 100+ lower case so going with lower case.

File: CODING_STANDARD.md (hunk 2 above)
Comment (thk123) [reply CODING_STANDARD.md]: Opted for:
> to make both the violation and the expected behavior clear

**Comment by thk123** [comment]
@danpoe @tautschnig I've had a go at updating the `invariant.h` file, let me know what you think.

File: src/util/invariant.h [hunk 3]
```diff
@@ -22,10 +22,18 @@ Author: Martin Brain, martin.brain@diffblue.com
 ** evaluate to a boolean.
 **
 ** As well as the condition they have a text argument that should be
-** used to say why they are true.  The reason should be a string literals.
+** used to say why they are true.  The reason should be a string literals
```
Comment (tautschnig) [inline src/util/invariant.h]: Nit pick: s/literals/literal/

**Review by tautschnig — APPROVED** [review]
Re-approving. Two wishlist items:
- As there now is some duplication of information, could you please add cross-referencing by mentioning invariant.h in CODING_STANDARD.md and vice-versa?
- The commit message "Update the invariant.cpp documentation to match" both seems to be incomplete (match what?) and refers to the wrong file (it's invariant.h).

**Comment by tautschnig** [comment]
@thk123 @danpoe as this has come up in #2963: could this PR please be revisited and merged ASAP?

**Review by chrisr-diffblue — APPROVED** [review]

**Review by danpoe — APPROVED** [review]

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: 43bfc739c34c9e4d2b53f02ca9762c277ee5fe49).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/86151309

**Pull request merged**